src/pwr.ps1

```diff
@@ -749,13 +749,17 @@ function Remove-Directory($Dir) {
 	}
 }
 
-function Get-InstalledPwrPackages {
+function Get-InstalledPwrPackages($Packages) {
+	$Names = if (-not $Packages -or $Packages.Count -eq 0) { @([regex]'^(.+)-([0-9._]+)$') } else { $Packages.ForEach{ if ($_.Replace('+', '_') -match '^([^:]+):([0-9._]+)$') { [regex]"^($($Matches[1]))-($($Matches[2].Replace('.', '\.'))(?:[._][0-9._]+)?)$" } else { [regex]"^($_)-([0-9._]+)$" } } }
```

Comment:
That's a line of code.


Can we just do this instead?

```ps1
function Get-InstalledPwrPackages {
	$Pkgs = @{}
	Get-ChildItem -Path $PwrPkgPath | ForEach-Object {
		if ($_.Name -match '(.+)-([0-9].+)') {
			$Pkg = $Matches[1]
			$Ver = $Matches[2]
			if (-not $Packages -or ($Packages | Where-Object {[regex]::Escape("${Pkg}:{$Ver}") -match $_})) {
				$Pkgs.$Pkg = @($Pkgs.$Pkg) + @([SemanticVersion]::new($Ver)) | Sort-Object -Descending
			}

		}
	}
	return $Pkgs
}
```

Reply:
Unfortunately, this allows things like `jdk:1` to match `jdk:17.0.1`, `cmake` to match `cmake-converter:1.2.0`, etc. However, I can make the long line less terrible.

Reply:
Oh, I kinda liked that. You don't, clearly. Were you looking for an exact match only, or no version specified?

Reply:
I don't hate it. It just seems weird to me to require it to match the Package to match the regex [name]:[version] (since it is part of the Packages parameter) and then match it to other things that are not that package or version. My intent was to match everything that falls into a specific package and/or version.

Reply:
Seems like we would want just some generic string, but not sure how to relax that parameter restriction for only `ls`.

BTW, here's the not-so-terrible version (untested), if you are interested in going that route:
```suggestion
	if (-not $Packages -or $Packages.Count -eq 0) {
		$Names = @([regex]'^(.+)-([0-9._]+)$')
	} else {
		$Names = @()
		foreach ($P in $Packages) {
			if ($_.Replace('+', '_') -match '^([^:]+):([0-9._]+)$') {
				$Names += [regex]"^($($Matches[1]))-($($Matches[2].Replace('.', '\.'))(?:[._][0-9._]+)?)$"
			} else {
				$Names += [regex]"^($_)-([0-9._]+)$"
			}
		}
	}
```

Reply:
Okay. I'll take another look at it

Reply:
I was playing with the behavior, and yours just feels a little rigid. Something about typing `vs` and seeing nothing made me sad. I think being more permissive when searching for packages and versions is a good thing. I was looking at other tools, and something like `winget` sort of behaves this way too.

<img width="343" alt="image" src="https://user-images.githubusercontent.com/58236504/200820431-5233f56a-9d76-4128-8c92-198af478e163.png">


Right now we both were using the `$Packages` to do the query, but I could see a case to add a string to the `-installed` param to add flexibility in query semantics.

For instance, if we could use strings like `:latest` to only match the latest version of all tools.

Reply:
Perhaps there could be flexibility enough to allow someone to have a query be an exact match only, kind of like what you wanted, but also be permissive in other scenarios like I want.

Reply:
Yeah, I'm fine with supporting this. However, I can't really use the `-Installed` parameter if I want to have the same functionality for `ls` without the `-Installed` parameter (e.g., `pwr ls vs, llvm`). Maybe, we just stick with the required pattern for `$Packages`, but change the regex for all matching packages? (The regex should be shorter as a result.)

Reply:
Here would be the new line to support this (assuming we don't switch from using the `$Packages` parameter):
```suggestion
	$Names = if (-not $Packages -or $Packages.Count -eq 0) { @([regex]'^(.+)-([0-9._]+)$') } else { $Packages.ForEach{ if ($_.Replace('+', '_') -match '^([^:]+):([0-9._]+)$') { [regex]"^(.*$($Matches[1]).*)-($($Matches[2].Replace('.', '\.'))[0-9._]*)$" } else { [regex]"^(.*$_.*)-([0-9._]+)$" } } }
```

It is one character shorter, but just as ugly if we keep it this way. What is your preference?
- [ ] Use this.
- [ ] Use this, but rewrite so there are more lines, but not as ugly.
- [ ] Rewrite how the `$Packages` parameter works (only check the regex when the command is not `ls`).
- [ ] Something else.

Reply:
Improved this to:
```suggestion
	$Names = if (-not $Packages -or $Packages.Count -eq 0) { @([regex]'^(.+)-([0-9._]+)$') } else { $Packages.ForEach{ [regex]($_.Replace('+', '_').Replace('.', '\.') -replace '^([^:]+)(?::([0-9._\\]+))?$', '^($1.*)-($2[0-9._]*)$') } }
```

Shorter, but still pretty ugly.

Reply:
So, I like this, but it seems to behave like the line of code I put earlier. Am I missing something? Wouldn't we prefer the simplicity of the that line over this?

Is this more easily applicable to the `ls` (non `-installed` version?

Reply:
They aren't exactly the same. The one you suggested will match things like `windows-terminal`, `perl` and `cmake-converter` for `er`, instead of just `erlang`. Whereas this one will only match packages starting with `er`. There is also a nuance with testing a version with a `+`, but that can be fixed pretty easily.

I'm honestly not sure if either of these solutions works well. If we end up with a package like `google-test` or , then it would basically be impossible to match only `go` without matching `google-test`. Which would be fine for displaying to a user, but my original intent was to get the exact version(s) of a package available, so I can load the same package with a different "config". (I can currently do this by "grepping" through the output of `pwr ls -Installed`, but was hoping to get something more precise.)

Maybe we could add something like a `-ExactMatch` option to handle this?

I think either one would be pretty easy to adapt to the non`-Installed` version.

Reply:
I updated this to include an `-ExactMatch` option. I also cleaned up some of the ugliness by adding a new function. The non-exact match does only match things that start with whatever is specified. If you want it to match just anything (e.g., `erl` matches `perl`), then I can change it.

Reply:
Oh. Okay. Then I feel like we went a bit overboard because I didn't understand your use case.

I'm fine with exact matches on the package and starts-with semantics for included versions with the package name. Revert what you had before.

Reply:
I can satisfy my itch in the future with a different command, something like `find` or `-find` maybe.

Reply:
Actually, since adding the `-ExactMatch` option, this does what I originally wanted. I'm fine with using that. The question I have is what should it do without `-ExactMatch`? Currently, I am using the "starts-with" semantics, but I can change it to be more like a "find" with "contains" semantics if you would prefer.

Or, I can just revert the `-ExactMatch` option and only exactly match, but that would be more work at this point. So, unless that is what you really want, we can just leave it in. Let me know.

Reply:
If you decide to keep the `-ExactMatch`, let me know, and I'll add documentation for it.

Reply:
I think `-ExactMatch` could be confusing to a user because it's only a name exact match.

Originally `list` was meant to show all packages and versions. If a _single_ package was given, it showed all the versions of that package.

This PR, conceptually, adds the ability to specify more than one package to just get the versions of those packages (_equal_ semantics) and additionally specify a version to those packages for _starts-with_ semantics.

I added an implementation for that behavior below, so you didn't have to rework anything.

I made the equals and starts-with behavior very explicit in my impl.

What do you think?

```diff
diff --git a/src/pwr.ps1 b/src/pwr.ps1
index 3660f66..714517e 100644
--- a/src/pwr.ps1
+++ b/src/pwr.ps1
@@ -761,6 +761,21 @@ function Get-InstalledPwrPackages {
 	return $Pkgs
 }
 
+function Select-PwrPackages([parameter(ValueFromPipeline=$True)]$Pkgs) {
+	if (-not $Packages) {
+		return $Pkgs
+	}
+	$Out = @{}
+	foreach ($Pkg in $Pkgs.Keys) {
+		foreach ($Ver in $Pkgs.$Pkg) {
+			if ($Packages | Where-Object { ($_ -match '^([^:]+):?(.*)$') -and ($Pkg -eq $Matches[1] -and ($Ver.ToString().StartsWith($Matches[2]))) } ) {
+				$Out.$Pkg = @($Out.$Pkg) + @([SemanticVersion]::new($Ver)) | Sort-Object -Descending
+			}
+		}
+	}
+	return $Out
+}
+
 function Format-PwrPackages([parameter(ValueFromPipeline=$True)]$Packages) {
 	''
 	$Packages | Format-List | Out-String -Stream | Where-Object{ $_.Length -gt 0 } | Sort-Object
@@ -968,7 +983,7 @@ function Invoke-PwrScripts {
 $ProgressPreference = 'SilentlyContinue'
 $PwrPath = if ($env:PwrHome) { $env:PwrHome } else { "$env:AppData\pwr" }
 $PwrPkgPath = "$PwrPath\pkg"
-$env:PwrVersion = '0.4.34'
+$env:PwrVersion = '0.4.35'
 Write-PwrInfo "running version $env:PwrVersion with powershell $($PSVersionTable.PSVersion)"
 
 if ($null -eq $Run) {
@@ -1170,24 +1185,21 @@ switch ($Command) {
 	}
 	{$_ -in 'ls', 'list'} {
 		if ($Installed) {
-			New-Object PSObject -Property (Get-InstalledPwrPackages) | Format-PwrPackages
-			break
-		}
-		foreach ($Repo in $PwrRepositories) {
-			if (($Packages.Count -eq 1) -and ($Packages[0] -match '[^:]+')) {
-				$Pkg = $Matches[0]
-				Write-PwrOutput "$Pkg[$($Repo.Uri)]"
-				if ($Repo.Packages.$Pkg -and -not $Quiet -and -not $Silent) {
-					$Repo.Packages.$Pkg | Format-List
-				} else {
-					Write-PwrHost '<none>'
-				}
+			Write-PwrOutput "[$env:PwrHome]"
+			$Show = Get-InstalledPwrPackages | Select-PwrPackages
+			if ($Show.Count -eq 0) {
+				Write-PwrHost '<none>'
 			} else {
+				New-Object PSObject -Property $Show | Format-PwrPackages
+			}
+		} else {
+			foreach ($Repo in $PwrRepositories) {
 				Write-PwrOutput "[$($Repo.Uri)]"
-				if ('' -ne $Repo.Packages -and -not $Quiet -and -not $Silent) {
-					$Repo.Packages | Format-PwrPackages
-				} else {
+				$Show = $Repo.Packages | ConvertTo-HashTable | Select-PwrPackages
+				if ($Show.Count -eq 0) {
 					Write-PwrHost '<none>'
+				} else {
+					New-Object PSObject -Property $Show | Format-PwrPackages
 				}
 			}
 		}

```

Reply:
Just tested this. Mostly, looks good to me.

I think this line: `Write-PwrOutput "[$env:PwrHome]"` should be `Write-PwrOutput "[$PwrPath]"`

The other thing I noticed is that `vs-buildtools:1` will match `vs-buildtools:17`. I don't think this matters for my use case, but it seems strange to me. It can be "fixed" by changing `$Ver.ToString().StartsWith($Matches[2])` to `-not $Matches[2] -or $Ver -match "^$([regex]::Escape($Matches[2]))(?:[^0-9]|$)"`

Reply:
Okay, I've updated this PR with what you have plus the 2 changes above. If you would rather still match using the start-with semantics, let me know, and I can change it back.